File: verification/acquire.ts
/**
 * REAL Tier 0 acquisition.
 *
 * Searches Earth Search for Sentinel-2 L2A scenes over the parcel for every
 * pre-registered window, reads the red, NIR and SCL bands over the sampling
 * frame via HTTP range requests, masks by SCL class, computes per-unit NDVI and
 * writes `verification/fixtures/tier0-<parcelId>.json`.
 *
 * Every scene ID and the processing-graph version are recorded so the
 * derivation can be re-run by a third party.
 *
 *   npm run acquire            # full run (network)
 *   npm run acquire -- --limit 5
 */
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { keccakOf } from './canonical.js';
import { readCogWindow, type CogWindowRead } from './cog.js';
import { acquisitionBboxUtm, buildSamplingFrame, type FrameUnit, type SamplingFrame } from './frame.js';
import { parcelIdentity } from './geometry.js';
import type { AnalysisPlan, ParcelRecord, SceneObservation, SceneRecord, Tier0Snapshot } from './models.js';
import { searchSentinel2, EARTH_SEARCH_URL, S2_L2A_COLLECTION } from './stac.js';

const here = dirname(fileURLToPath(import.meta.url));
export const FIXTURES_DIR = join(here, 'fixtures');
const CACHE_DIR = join(here, 'cache');

export const PROCESSING_GRAPH_VERSION = '1.0.0';

export interface SceneBands {
  red: CogWindowRead;
  nir: CogWindowRead;
  scl: CogWindowRead;
}

/**
 * Per-unit NDVI statistics for one scene. Pure: given the three band windows
 * and the frame, returns the observation row. Exported for unit tests.
 */
export function observeScene(scene: SceneRecord, bands: SceneBands, units: FrameUnit[], plan: AnalysisPlan): SceneObservation {
  const { red, nir, scl } = bands;
  if (red.width !== nir.width || red.height !== nir.height) throw new Error(`${scene.sceneId}: red/nir window mismatch`);
  const validClasses = new Set(plan.index.sclClassesValid);
  const { scale, offset } = plan.index.dnToReflectance;
  const g = red.grid;
  const ndvi = new Float32Array(red.width * red.height);
  const valid = new Uint8Array(red.width * red.height);
  const water = new Uint8Array(red.width * red.height);
  for (let r = 0; r < red.height; r++) {
    const y = g.originY - (r + 0.5) * g.resolution;
    const sr = Math.floor((scl.grid.originY - y) / scl.grid.resolution);
    for (let c = 0; c < red.width; c++) {
      const i = r * red.width + c;
      const x = g.originX + (c + 0.5) * g.resolution;
      const sc = Math.floor((x - scl.grid.originX) / scl.grid.resolution);
      const sclVal = sr >= 0 && sr < scl.height && sc >= 0 && sc < scl.width ? scl.data[sr * scl.width + sc]! : 0;
      const dnR = red.data[i]!;
      const dnN = nir.data[i]!;
      if (dnR === 0 || dnN === 0 || !validClasses.has(sclVal)) continue;
      const rr = dnR * scale + offset;
      const nn = dnN * scale + offset;
      const denom = nn + rr;
      if (denom <= 0) continue;
      const v = (nn - rr) / denom;
      if (v < -1 || v > 1) continue;
      ndvi[i] = v;
      valid[i] = 1;
      if (sclVal === 6) water[i] = 1;
    }
  }
  const out: SceneObservation = { ndvi: [], validFraction: [], waterFraction: [] };
  for (const u of units) {
    let n = 0;
    let s = 0;
    let w = 0;
    for (const px of u.pixels) {
      if (!valid[px]) continue;
      n++;
      s += ndvi[px]!;
      w += water[px]!;
    }
    const vf = u.pixels.length > 0 ? n / u.pixels.length : 0;
    out.validFraction.push(round4(vf));
    out.waterFraction.push(n > 0 ? round4(w / n) : 0);
    out.ndvi.push(vf >= plan.index.minValidPixelFraction && n > 0 ? round4(s / n) : null);
  }
  return out;
}

function round4(x: number): number {
  return Math.round(x * 10_000) / 10_000;
}

export function loadParcel(): ParcelRecord {
  return JSON.parse(readFileSync(join(FIXTURES_DIR, 'kootenay-parcel.json'), 'utf8')) as ParcelRecord;
}

export function loadPlan(): AnalysisPlan {
  return JSON.parse(readFileSync(join(FIXTURES_DIR, 'analysis-plan.json'), 'utf8')) as AnalysisPlan;
}

export function finalizeSnapshot(partial: Omit<Tier0Snapshot, 'snapshotHash'>): Tier0Snapshot {
  return { ...partial, snapshotHash: keccakOf(partial) };
}

async function readBands(scene: SceneRecord, bboxUtm: [number, number, number, number]): Promise<SceneBands> {
  const [red, nir, scl] = await Promise.all([
    readCogWindow(scene.assets.red, { bboxUtm, cacheDir: CACHE_DIR }),
    readCogWindow(scene.assets.nir, { bboxUtm, cacheDir: CACHE_DIR }),
    readCogWindow(scene.assets.scl, { bboxUtm, cacheDir: CACHE_DIR }),
  ]);
  return { red, nir, scl };
}

async function mapLimit<T, R>(items: T[], limit: number, fn: (item: T, i: number) => Promise<R>): Promise<R[]> {
  const out: R[] = new Array<R>(items.length);
  let next = 0;
  const workers = Array.from({ length: Math.min(limit, items.length) }, async () => {
    while (next < items.length) {
      const i = next++;
      out[i] = await fn(items[i]!, i);
    }
  });
  await Promise.all(workers);
  return out;
}

export async function acquire(opts: { limit?: number; concurrency?: number; log?: (s: string) => void } = {}): Promise<Tier0Snapshot> {
  const log = opts.log ?? ((s: string) => console.log(s));
  const parcel = loadParcel();
  const plan = loadPlan();
  const identity = parcelIdentity(parcel.parcelId, parcel.geometry, parcel.h3Resolution);
  const bboxUtm = acquisitionBboxUtm(parcel, plan);
  const [minLng, minLat, maxLng, maxLat] = bboxLngLat(parcel);
  log(`parcel ${parcel.parcelId}: ${identity.areaHa.toFixed(2)} ha, ${identity.h3CellCount} H3 r${parcel.h3Resolution} cells, geometryHash ${identity.geometryHash}`);

  let scenes = await searchSentinel2({
    bbox: [minLng, minLat, maxLng, maxLat],
    windows: [...plan.windows.pre, ...plan.windows.post],
    maxCloudCoverPct: plan.index.maxSceneCloudCoverPct,
  });
  if (opts.limit) scenes = scenes.slice(0, opts.limit);
  log(`STAC: ${scenes.length} Sentinel-2 L2A scenes under ${plan.index.maxSceneCloudCoverPct}% cloud`);

  // First scene defines the grid; every other scene must match it (same MGRS tile).
  const first = await readBands(scenes[0]!, bboxUtm);
  const grid = first.red.grid;
  let frame: SamplingFrame = buildSamplingFrame(parcel, plan, grid);
  log(`frame: ${frame.units.length} units (${count(frame, 'parcel_cell')} parcel cells, ${count(frame, 'near')} near, ${count(frame, 'far')} far) on a ${grid.width}x${grid.height} px grid`);

  const observations: Record<string, SceneObservation> = {};
  const kept: SceneRecord[] = [];
  const results = await mapLimit(scenes, opts.concurrency ?? 4, async (scene, i) => {
    const bands = i === 0 ? first : await readBands(scene, bboxUtm);
    const g = bands.red.grid;
    if (g.originX !== grid.originX || g.originY !== grid.originY || g.width !== grid.width || g.height !== grid.height) {
      log(`  skip ${scene.sceneId}: grid mismatch`);
      return null;
    }
    const obs = observeScene(scene, bands, frame.units, plan);
    log(`  ${scene.sceneId} ${scene.datetime.slice(0, 10)} cloud ${scene.cloudCoverPct.toFixed(1)}% parcel NDVI ${obs.ndvi[0] ?? 'masked'} (valid ${obs.validFraction[0]})`);
    return { scene, obs };
  });
  for (const r of results) {
    if (!r) continue;
    kept.push(r.scene);
    observations[r.scene.sceneId] = r.obs;
  }
  frame = { ...frame, units: frame.units };

  const snapshot = finalizeSnapshot({
    provenance: 'REAL',
    tier: 0,
    source: { catalog: EARTH_SEARCH_URL, collection: S2_L2A_COLLECTION, assetHost: 'sentinel-cogs.s3.us-west-2.amazonaws.com' },
    processingGraphVersion: PROCESSING_GRAPH_VERSION,
    acquiredAt: new Date().toISOString(),
    parcelId: parcel.parcelId,
    geometryHash: identity.geometryHash,
    h3Root: identity.h3Root,
    crs: 'EPSG:32611',
    window: { bbox: bboxUtm, pixelWindow: [0, 0, grid.width, grid.height] },
    scenes: kept,
    units: frame.units.map(({ pixels: _pixels, ...u }) => u),
    observations,
  });
  mkdirSync(FIXTURES_DIR, { recursive: true });
  const outPath = join(FIXTURES_DIR, `tier0-${parcel.parcelId}.json`);
  writeFileSync(outPath, JSON.stringify(snapshot));
  log(`wrote ${outPath} (${kept.length} scenes, snapshotHash ${snapshot.snapshotHash})`);
  return snapshot;
}

function count(frame: SamplingFrame, zone: string): number {
  return frame.units.filter((u) => u.zone === zone).length;
}

function bboxLngLat(parcel: ParcelRecord): [number, number, number, number] {
  const ring = parcel.geometry.coordinates[0]!;
  const lngs = ring.map((p) => p[0]!);
  const lats = ring.map((p) => p[1]!);
  return [Math.min(...lngs), Math.min(...lats), Math.max(...lngs), Math.max(...lats)];
}

const isMain = process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1];
if (isMain) {
  const limitArg = process.argv.indexOf('--limit');
  const limit = limitArg > 0 ? Number(process.argv[limitArg + 1]) : undefined;
  acquire(limit ? { limit } : {}).catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: analysis/ecorestore_analysis/acquire.py
"""REAL Tier 0 acquisition — the batch job (processing graph 2.0.0).

Searches Earth Search for Sentinel-2 L2A scenes over the parcel for every
pre-registered window, reads the red, NIR and SCL bands over the sampling
frame through rasterio, masks by SCL class, computes per-unit NDVI and writes
an **unhashed** snapshot:

    <out>/tier0-<parcelId>.unhashed.json

It is the same document as ``verification/fixtures/tier0-<parcelId>.json``
minus ``geometryHash``, ``h3Root`` and ``snapshotHash``. Those are attached
by ``npm run acquire:finalize`` — the TypeScript canonicaliser — which then
writes the fixture. Python produces observations; it commits nothing.

    ecorestore-acquire --out ./out/acquire --limit 5
    ecorestore-acquire --fixtures ../verification/fixtures --out ./out/acquire
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import numpy as np

from . import PROCESSING_GRAPH_VERSION
from .cog import CogWindowRead, read_cog_window
from .geometry import FrameUnit, PixelGrid, acquisition_bbox_utm, build_sampling_frame, utm_epsg_for
from .jsnum import round4
from .stac import ASSET_HOST, EARTH_SEARCH_URL, S2_L2A_COLLECTION, search_sentinel2


def observe_scene(red: CogWindowRead, nir: CogWindowRead, scl: CogWindowRead, units: list[FrameUnit], plan: dict[str, Any]) -> dict[str, list]:
    """Per-unit NDVI statistics for one scene (``observeScene`` in ``acquire.ts``)."""
    if red.width != nir.width or red.height != nir.height:
        raise ValueError("red/nir window mismatch")
    valid_classes = np.asarray(plan["index"]["sclClassesValid"])
    scale = plan["index"]["dnToReflectance"]["scale"]
    offset = plan["index"]["dnToReflectance"]["offset"]
    g = red.grid
    rows = np.arange(red.height)
    cols = np.arange(red.width)
    y = g.origin_y - (rows + 0.5) * g.resolution
    x = g.origin_x + (cols + 0.5) * g.resolution
    sr = np.floor((scl.grid.origin_y - y) / scl.grid.resolution).astype(np.int64)
    sc = np.floor((x - scl.grid.origin_x) / scl.grid.resolution).astype(np.int64)
    SR, SC = np.meshgrid(sr, sc, indexing="ij")
    in_range = (SR >= 0) & (SR < scl.height) & (SC >= 0) & (SC < scl.width)
    scl_val = np.zeros((red.height, red.width), dtype=np.int64)
    scl_val[in_range] = scl.data[SR[in_range], SC[in_range]]

    dn_r = red.data.astype(np.float64)
    dn_n = nir.data.astype(np.float64)
    valid = (dn_r != 0) & (dn_n != 0) & np.isin(scl_val, valid_classes)
    rr = dn_r * scale + offset
    nn = dn_n * scale + offset
    denom = nn + rr
    with np.errstate(divide="ignore", invalid="ignore"):
        v = np.where(denom > 0, (nn - rr) / denom, np.nan)
    valid &= denom > 0
    valid &= (v >= -1) & (v <= 1)
    # The reference stores NDVI in a Float32Array and accumulates in float64.
    ndvi = np.where(valid, v, 0.0).astype(np.float32).astype(np.float64).ravel()
    valid_flat = valid.ravel()
    water_flat = (valid & (scl_val == 6)).ravel()

    out: dict[str, list] = {"ndvi": [], "validFraction": [], "waterFraction": []}
    min_valid = plan["index"]["minValidPixelFraction"]
    for u in units:
        px = u.pixels
        vmask = valid_flat[px]
        n = int(vmask.sum())
        s = float(ndvi[px][vmask].sum())
        w = int(water_flat[px].sum())
        vf = n / len(px) if len(px) > 0 else 0.0
        out["validFraction"].append(round4(vf))
        out["waterFraction"].append(round4(w / n) if n > 0 else 0)
        out["ndvi"].append(round4(s / n) if vf >= min_valid and n > 0 else None)
    return out


def load_fixture(fixtures: Path, name: str) -> dict[str, Any]:
    return json.loads((fixtures / name).read_text())


def acquire(fixtures: Path, out_dir: Path, cache_dir: Path | None, limit: int | None, concurrency: int, log=print) -> Path:
    parcel = load_fixture(fixtures, "kootenay-parcel.json")
    plan = load_fixture(fixtures, "analysis-plan.json")
    ring = parcel["geometry"]["coordinates"][0]
    lngs = [p[0] for p in ring]
    lats = [p[1] for p in ring]
    bbox = (min(lngs), min(lats), max(lngs), max(lats))

    scenes = search_sentinel2(bbox, [*plan["windows"]["pre"], *plan["windows"]["post"]], plan["index"]["maxSceneCloudCoverPct"])
    if limit:
        scenes = scenes[:limit]
    log(f"STAC: {len(scenes)} Sentinel-2 L2A scenes under {plan['index']['maxSceneCloudCoverPct']}% cloud")
    if not scenes:
        raise SystemExit("no scenes found")

    # The frame is built in the UTM zone of the tile being read. The first scene
    # defines the grid, so its zone defines the frame; a scene from another zone
    # can never share the grid and is skipped before any band is downloaded.
    epsg = scenes[0]["epsg"] or utm_epsg_for((bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2)
    bbox_utm = acquisition_bbox_utm(parcel["geometry"], plan, epsg)
    log(f"parcel {parcel['parcelId']}: EPSG:{epsg}, acquisition bbox {tuple(round(v, 1) for v in bbox_utm)}")

    def read_bands(scene: dict[str, Any]) -> tuple[CogWindowRead, CogWindowRead, CogWindowRead]:
        a = scene["assets"]
        return (read_cog_window(a["red"], bbox_utm, cache_dir), read_cog_window(a["nir"], bbox_utm, cache_dir), read_cog_window(a["scl"], bbox_utm, cache_dir))

    # First scene defines the grid; every other scene must match it (same MGRS tile).
    first = read_bands(scenes[0])
    grid: PixelGrid = first[0].grid
    units = build_sampling_frame(parcel["geometry"], plan, grid, epsg)
    counts = {z: sum(1 for u in units if u.zone == z) for z in ("parcel_cell", "near", "far")}
    log(f"frame: {len(units)} units ({counts['parcel_cell']} parcel cells, {counts['near']} near, {counts['far']} far) on a {grid.width}x{grid.height} px grid")

    observations: dict[str, Any] = {}
    kept: list[dict[str, Any]] = []

    def work(i_scene: tuple[int, dict[str, Any]]):
        i, scene = i_scene
        if scene["epsg"] and scene["epsg"] != epsg:
            log(f"  skip {scene['sceneId']}: EPSG:{scene['epsg']}, frame is EPSG:{epsg}")
            return None
        bands = first if i == 0 else read_bands(scene)
        g = bands[0].grid
        if (g.origin_x, g.origin_y, g.width, g.height) != (grid.origin_x, grid.origin_y, grid.width, grid.height):
            log(f"  skip {scene['sceneId']}: grid mismatch")
            return None
        obs = observe_scene(*bands, units, plan)
        log(f"  {scene['sceneId']} {scene['datetime'][:10]} cloud {scene['cloudCoverPct']:.1f}% parcel NDVI {obs['ndvi'][0] if obs['ndvi'][0] is not None else 'masked'} (valid {obs['validFraction'][0]})")
        return scene, obs

    with ThreadPoolExecutor(max_workers=concurrency) as pool:
        for r in pool.map(work, enumerate(scenes)):
            if r is None:
                continue
            scene, obs = r
            kept.append(scene)
            observations[scene["sceneId"]] = obs

    snapshot = {
        "provenance": "REAL",
        "tier": 0,
        "source": {"catalog": EARTH_SEARCH_URL, "collection": S2_L2A_COLLECTION, "assetHost": ASSET_HOST},
        "processingGraphVersion": PROCESSING_GRAPH_VERSION,
        "acquiredAt": datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
        "parcelId": parcel["parcelId"],
        "crs": f"EPSG:{epsg}",
        "window": {"bbox": list(bbox_utm), "pixelWindow": [0, 0, grid.width, grid.height], "grid": grid.to_dict()},
        "scenes": kept,
        "units": [u.to_dict() for u in units],
        "observations": observations,
        # Handoff only — `scripts/finalize-acquisition.ts` verifies this against
        # the repository parcel before attaching `geometryHash` and `h3Root`,
        # then strips it. It is never part of the hashed document. Without it
        # finalize could bind a snapshot acquired from a different `--fixtures`
        # geometry to the repository parcel's identity on a matching id alone.
        "sourceParcel": {
            "parcelId": parcel["parcelId"],
            "geometry": parcel["geometry"],
            "h3Resolution": parcel["h3Resolution"],
        },
    }
    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / f"tier0-{parcel['parcelId']}.unhashed.json"
    out_path.write_text(json.dumps(snapshot, allow_nan=False))
    log(f"wrote {out_path} ({len(kept)} scenes, unhashed — run `npm run acquire:finalize -- {out_path}` to attach geometryHash, h3Root and snapshotHash)")
    return out_path


def main(argv: list[str] | None = None) -> None:
    ap = argparse.ArgumentParser(description="REAL Tier 0 acquisition (processing graph 2.0.0)")
    ap.add_argument("--fixtures", default=None, help="directory holding kootenay-parcel.json and analysis-plan.json (default: /fixtures, else ../verification/fixtures)")
    ap.add_argument("--out", default=os.environ.get("ACQUIRE_OUT", "./out/acquire"), help="output directory (env ACQUIRE_OUT)")
    ap.add_argument("--cache", default=os.environ.get("ACQUIRE_CACHE_DIR"), help="COG window cache directory (env ACQUIRE_CACHE_DIR)")
    ap.add_argument("--limit", type=int, default=None, help="only the first N scenes (smoke test)")
    ap.add_argument("--concurrency", type=int, default=4)
    args = ap.parse_args(argv)
    fixtures = Path(args.fixtures) if args.fixtures else (Path("/fixtures") if Path("/fixtures").is_dir() else Path(__file__).resolve().parents[2] / "verification" / "fixtures")
    if not (fixtures / "analysis-plan.json").exists():
        sys.exit(f"fixtures not found at {fixtures}")
    t0 = time.perf_counter()
    acquire(fixtures, Path(args.out), Path(args.cache) if args.cache else None, args.limit, args.concurrency)
    print(f"done in {time.perf_counter() - t0:.1f} s")


if __name__ == "__main__":
    main()
